py/objtype.c

```diff
@@ -587,17 +587,16 @@ STATIC void mp_obj_instance_load_attr(mp_obj_t self_in, qstr attr, mp_obj_t *des
     }
     #if MICROPY_CPYTHON_COMPAT
     if (attr == MP_QSTR___dict__) {
-        // Create a new dict with a copy of the instance's map items.
+        // Returns a read-only dict of the instance members.
+        // If the internal locals is not fixed, a copy will be created.
         // This creates, unlike CPython, a 'read-only' __dict__: modifying
         // it will not result in modifications to the actual instance members.
-        mp_map_t *map = &self->members;
-        mp_obj_t attr_dict = mp_obj_new_dict(map->used);
-        for (size_t i = 0; i < map->alloc; ++i) {
-            if (mp_map_slot_is_filled(map, i)) {
-                mp_obj_dict_store(attr_dict, map->table[i].key, map->table[i].value);
-            }
-        }
-        dest[0] = attr_dict;
+        mp_obj_dict_t dict;
+        dict.map = self->members;
+        dict.base.type = &mp_type_dict;
+        dest[0] = mp_obj_dict_copy(MP_OBJ_FROM_PTR(&dict));
+        mp_obj_dict_t *dest_dict = MP_OBJ_TO_PTR(dest[0]);
+        dest_dict->map.is_fixed = 1;
```

Comment:
Can you please put these changes to the instance `__dict__` lookup into a separate commit?

@stinos as the original author of this code, what are your thoughts on this change?  In particular do you think it's a good idea to set `is_fixed=1` to prevent modification of the returned dict (and so prevent subtle errors when the user tries to change the dict and expects those changes to be mirrored in the instance, which they are not)?

Reply:
Would be my preference to reduce chances of things being misused yes. And looking at the original discussion in https://github.com/micropython/micropython/pull/1757, I actually wanted it to be like that but that wasn't possible back then.

Reply:
Ok, so let's keep the `is_fixed=1` part, to make the returned dict R/O.  It will be a breaking change though, if users are relying on the dict being modifiable (eg because they know it's not reflected in the instance but they want anyway to reuse the dict for something else).  But it'll fail pretty quickly now if someone is doing this and they can fix their code by simply doing `d = dict(my_type.__dict__)`